README.md
# RPGMakerMZ
RPGツクールMZ用のプラグインです。


docs/HISTORY.md
# 変更履歴
ここでは各プラグインの変更履歴を記載しています。  
一番上が最新です。

### Sbm_KeyPatchForBothLR.js
 - 1.0.0 2021/06/12 Sbm_ReversibleKeyPatch.jsより改称
### ~~Sbm_ReversibleKeyPatch.js~~
- プラグイン名変更に伴い廃止
- 1.0.0 2021/06/05 初版
### Sbm_KeyPatchForNumpad.js
 - 1.0.0 2021/06/12 Sbm_NumericKeyPatch.jsより改称
### ~~Sbm_NumericKeyPatch.js~~
- プラグイン名変更に伴い廃止
- 1.0.1 2021/06/08
    - 一時オブジェクトを単一化
    - ヘルプの加筆修正
- 1.0.0 2021/06/06 初版
### Sbm_KeyPatchForGamepad.js
 - 1.0.0 2021/06/12 Sbm_GamingKeyPatch.jsより改称
### ~~Sbm_GamingKeyPatch.js~~
 - プラグイン名変更に伴い廃止
 - 1.0.0 2021/06/08 初版


Sbm_GamingKeyPatch.js
/*
===============================================================================
 RPG Maker MZ - Sbm GamingKeyPatch (1.0.0)
===============================================================================
 Copyright (c) 2021 Suib
 This software is released under the MIT License.
 http://www.opensource.org/licenses/mit-license
===============================================================================
*/

/*:
 * @target MZ
 * @plugindesc This plugin assigns any key settings to the Gamepad.
 * @author Suib
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_GamingKeyPatch.js
 * @base PluginCommonBase
 *
 * @param keyMapArrButton
 * @text keySettings:Button
 * @desc Assign operations to buttons (multiple can be selected).
 * @type struct<KeyMapButton>[]
 * @default []
 * 
 * @param keyMapArrDpad
 * @text keySettings:D-pad
 * @desc Assign operations to D-pad (multiple can be selected).
 * @type struct<KeyMapDpad>[]
 * @default []
 * 
 * @help Sbm_GamingKeyPatch.js
 *
 * -- Summary --
 * This plugin assigns any key settings to the gamepad.
 * Unassigned keys retain their default key settings.
 * The key layout is for the Xbox controller.
 * 
 * -- Usage --
 * Select the setting target from "Button" or "D-pad" in the plugin parameters.
 * Select any "keycode" in the pull-down item and select the "operation" to be assigned to it.
 * 
 * -- Plugin Parameters --
 * keySettings:Button (multiple can be selected)
 * keySettings:D-pad (multiple can be selected)
 * 
 * -- Base Plugins --
 * PluginCommonBase
 * 
 * -- Terms of Use --
 * Comply with the MIT license agreement.
 * http://www.opensource.org/licenses/mit-license
 * 
 */
/*~struct~KeyMapButton:
 *
 * @param keyCode
 * @text keyCode
 * @desc Select the button to which you want to assign the operation.
 * @type select
 * @option A
 * @value 0
 * @option B
 * @value 1
 * @option X
 * @value 2
 * @option Y
 * @value 3
 * @option LB
 * @value 4
 * @option RB
 * @value 5
 * @option LT
 * @value 6
 * @option RT
 * @value 7
 * @option Back
 * @value 8
 * @option Start
 * @value 9
 * @option LS-Press
 * @value 10
 * @option RS-Press
 * @value 11
 * 
 * @param operation
 * @text operation
 * @desc Select the operation to assign to the target button.
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */
/*~struct~KeyMapDpad:
 *
 * @param keyCode
 * @text keyCode
 * @desc Select the directional key to which you want to assign the operation.
 * @type select
 * @option up
 * @value 12
 * @option down
 * @value 13
 * @option left
 * @value 14
 * @option right
 * @value 15
 * 
 * @param operation
 * @text operation
 * @desc Select the operation to assign to the target directional key.
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */

/*:ja
 * @target MZ
 * @plugindesc このプラグインは、ゲームパッドに任意のキー設定を割り当てます。
 * @author 水部
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_GamingKeyPatch.js
 * @base PluginCommonBase
 *
 * @param keyMapArrButton
 * @text キー設定：ボタン(複数選択可)
 * @desc 各ボタンに任意の操作を割り当てます(複数選択可)。
 * @type struct<KeyMapButton>[]
 * @default []
 * 
 * @param keyMapArrDpad
 * @text キー設定：方向キー(複数選択可)
 * @desc 方向キーに任意の操作を割り当てます(複数選択可)。
 * @type struct<KeyMapDpad>[]
 * @default []
 * 
 * @help Sbm_GamingKeyPatch.js
 *
 * 【概要】
 * このプラグインは、ゲームパッドの各ボタンに任意のキー設定を割り当てます。
 * 割り当てが行われなかったキーはデフォルトのキー設定を保持します。
 * キーレイアウトはXboxコントローラーを想定しています。
 * 
 * 【使用方法】
 * ・プラグインパラメータで、設定対象のキーを「ボタン」または「方向キー」から選択します。
 * ・プルダウン項目で任意の「キー」を選択し、それに割り当てる「操作」を選択します。
 * 
 * 【プラグインパラメータ】
 * ・キー設定：ボタン(複数選択可)
 * ・キー設定：方向キー(複数選択可)
 * 
 * 【前提プラグイン】
 * ・PluginCommonBase
 * 
 * 【利用規約】
 * MITライセンス規約に準じます。
 * https://licenses.opensource.jp/MIT/MIT
 * 
 */
/*~struct~KeyMapButton:ja
 *
 * @param keyCode
 * @text キーコード
 * @desc 操作を割り当てる対象のボタンを選択します。
 * @type select
 * @option A
 * @value 0
 * @option B
 * @value 1
 * @option X
 * @value 2
 * @option Y
 * @value 3
 * @option LB
 * @value 4
 * @option RB
 * @value 5
 * @option LT
 * @value 6
 * @option RT
 * @value 7
 * @option Back
 * @value 8
 * @option Start
 * @value 9
 * @option LS-Press
 * @value 10
 * @option RS-Press
 * @value 11
 * 
 * @param operation
 * @text 操作
 * @desc 対象のボタンに割り当てる操作を選択します。
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */
/*~struct~KeyMapDpad:ja
 *
 * @param keyCode
 * @text キーコード
 * @desc 操作を割り当てる対象の方向キーを選択します。
 * @type select
 * @option up
 * @value 12
 * @option down
 * @value 13
 * @option left
 * @value 14
 * @option right
 * @value 15
 * 
 * @param operation
 * @text 操作
 * @desc 対象の方向キーに割り当てる操作を選択します。
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */

(() => {
    'use strict';
    const script = document.currentScript;
    const param = PluginManagerEx.createParameter(script);

    const keyMapGamepad = {};
    if (Array.isArray(param.keyMapArrButton)) {
        for (const keyMap of param.keyMapArrButton) {
            Object.assign(keyMapGamepad, { [String(keyMap.keyCode)]: keyMap.operation });
        }
    }
    if (Array.isArray(param.keyMapArrDpad)) {
        for (const keyMap of param.keyMapArrDpad) {
            Object.assign(keyMapGamepad, { [String(keyMap.keyCode)]: keyMap.operation });
        }
    }
    // ==============================
    // Input
    // ==============================
    Object.assign(Input.gamepadMapper, keyMapGamepad);
})();


Sbm_NumericKeyPatch.js
/*
===============================================================================
 RPG Maker MZ - Sbm NumericKeyPatch (1.0.1)
===============================================================================
 Copyright (c) 2021 Suib
 This software is released under the MIT License.
 http://www.opensource.org/licenses/mit-license
===============================================================================
*/

/*:
 * @target MZ
 * @plugindesc This plugin assigns any key settings to the Numpad.
 * @author Suib
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_NumericKeyPatch.js
 * @base PluginCommonBase
 * @orderAfter Sbm_ReversibleKeyPatch
 *
 * @param keyMapArrNumber
 * @text keySettings:Number
 * @desc Assign operations to the key to "0-9" (multiple can be selected).
 * @type struct<KeyMapNumber>[]
 * @default []
 * 
 * @param keyMapArrSymbol
 * @text keySettings:Symbol
 * @desc Assign operations to keys such as "+" and "-" (multiple can be selected).
 * @type struct<KeyMapSymbol>[]
 * @default []
 * 
 * @help Sbm_NumericKeyPatch.js
 *
 * -- Summary --
 * This plugin assigns any key settings to the numeric keypad.
 * Unassigned keys retain their default key settings.
 * 
 * -- Usage --
 * Select the setting target from "Number" or "Symbol" in the plugin parameters.
 * Select any "keycode" in the pull-down item and select the "operation" to be assigned to it.
 * 
 * -- Plugin Parameters --
 * keySettings:Number (multiple can be selected)
 * keySettings:Symbol (multiple can be selected)
 * 
 * -- Base Plugins --
 * PluginCommonBase
 * 
 * -- Terms of Use --
 * Comply with the MIT license agreement.
 * http://www.opensource.org/licenses/mit-license
 * 
 */
/*~struct~KeyMapNumber:
 *
 * @param keyCode
 * @text keyCode
 * @desc Select the key to which you want to assign the operation.
 * @type select
 * @option Numpad 0
 * @value 96
 * @option Numpad 1
 * @value 97
 * @option Numpad 2
 * @value 98
 * @option Numpad 3
 * @value 99
 * @option Numpad 4
 * @value 100
 * @option Numpad 5
 * @value 101
 * @option Numpad 6
 * @value 102
 * @option Numpad 7
 * @value 103
 * @option Numpad 8
 * @value 104
 * @option Numpad 9
 * @value 105
 * 
 * @param operation
 * @text operation
 * @desc Select the operation to assign to the target key.
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */
/*~struct~KeyMapSymbol:
 *
 * @param keyCode
 * @text keyCode
 * @desc Select the key to which you want to assign the operation.
 * @type select
 * @option Numpad +
 * @value 107
 * @option Numpad -
 * @value 109
 * @option Numpad *
 * @value 106
 * @option Numpad /
 * @value 111
 * @option Numpad .
 * @value 110
 * 
 * @param operation
 * @text operation
 * @desc Select the operation to assign to the target key.
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */

/*:ja
 * @target MZ
 * @plugindesc このプラグインは、テンキーに任意のキー設定を割り当てます。
 * @author 水部
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_NumericKeyPatch.js
 * @base PluginCommonBase
 * @orderAfter Sbm_ReversibleKeyPatch
 *
 * @param keyMapArrNumber
 * @text キー設定：数字(複数選択可)
 * @desc 「0-9」までのキーに操作を割り当てます(複数選択可)。
 * @type struct<KeyMapNumber>[]
 * @default []
 * 
 * @param keyMapArrSymbol
 * @text キー設定：記号(複数選択可)
 * @desc 「+」や「-」等のキーに操作を割り当てます(複数選択可)。
 * @type struct<KeyMapSymbol>[]
 * @default []
 * 
 * @help Sbm_NumericKeyPatch.js
 *
 * 【概要】
 * このプラグインは、テンキーに任意のキー設定を割り当てます。
 * 割り当てが行われなかったキーはデフォルトのキー設定を保持します。
 * 
 * 【使用方法】
 * ・プラグインパラメータで、設定対象のキーを「数字」または「記号」から選択します。
 * ・プルダウン項目で任意の「キー」を選択し、それに割り当てる「操作」を選択します。
 * 
 * 【プラグインパラメータ】
 * ・キー設定：数字(複数選択可)
 * ・キー設定：記号(複数選択可)
 * 
 * 【前提プラグイン】
 * ・PluginCommonBase
 * 
 * 【利用規約】
 * MITライセンス規約に準じます。
 * https://licenses.opensource.jp/MIT/MIT
 * 
 */
/*~struct~KeyMapNumber:ja
 *
 * @param keyCode
 * @text キーコード
 * @desc 操作を割り当てる対象のキーを選択します。
 * @type select
 * @option Numpad 0
 * @value 96
 * @option Numpad 1
 * @value 97
 * @option Numpad 2
 * @value 98
 * @option Numpad 3
 * @value 99
 * @option Numpad 4
 * @value 100
 * @option Numpad 5
 * @value 101
 * @option Numpad 6
 * @value 102
 * @option Numpad 7
 * @value 103
 * @option Numpad 8
 * @value 104
 * @option Numpad 9
 * @value 105
 * 
 * @param operation
 * @text 操作
 * @desc 対象のキーに割り当てる操作を選択します。
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */
/*~struct~KeyMapSymbol:ja
 *
 * @param keyCode
 * @text キーコード
 * @desc 操作を割り当てる対象のキーを選択します。
 * @type select
 * @option Numpad +
 * @value 107
 * @option Numpad -
 * @value 109
 * @option Numpad *
 * @value 106
 * @option Numpad /
 * @value 111
 * @option Numpad .
 * @value 110
 * 
 * @param operation
 * @text 操作
 * @desc 対象のキーに割り当てる操作を選択します。
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */

(() => {
    'use strict';
    const script = document.currentScript;
    const param = PluginManagerEx.createParameter(script);

    const keyMapNumpad = {};
    if (Array.isArray(param.keyMapArrNumber)) {
        for (const keyMap of param.keyMapArrNumber) {
            Object.assign(keyMapNumpad, { [String(keyMap.keyCode)]: keyMap.operation });
        }
    }
    if (Array.isArray(param.keyMapArrSymbol)) {
        for (const keyMap of param.keyMapArrSymbol) {
            Object.assign(keyMapNumpad, { [String(keyMap.keyCode)]: keyMap.operation });
        }
    }
    // ==============================
    // Input
    // ==============================
    Object.assign(Input.keyMapper, keyMapNumpad);
})();


Sbm_KeyPatchForBothLR.js
/*
===============================================================================
 RPG Maker MZ - Sbm KeyPatchForBothLR (1.0.0)
===============================================================================
 Copyright (c) 2021 Suib
 This software is released under the MIT License.
 http://www.opensource.org/licenses/mit-license
===============================================================================
*/

/*:
 * @target MZ
 * @plugindesc this plugin provides key patch that For both left and right uses.
 * @author Suib
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_KeyPatchForBothLR.js
 *
 * @help Sbm_KeyPatchForBothLR.js
 *
 * -- Summary --
 * this plugin provides key patch that For both left and right uses.
 * 
 * -- Usage --
 * It is applied only by installing this plugin.
 * 
 * -- Changes --
 * <Keyboard>
 * - W          : pagedown  -> up arrow
 * - Alt        : control   -> alt
 * <NumKeypad>
 * - 0          : escape    -> shift
 * - 2          : down      -> cancel
 * - 4          : left      -> pageup
 * - 6          : right     -> pagedown
 * - 8          : up        -> tab
 * 
 * -- Additions --
 * <Keyboard>
 * - W/A/S/D    : up/left/down/right *(arrow)
 * - Q/E        : pageup/pagedown
 * - C          : ok
 * - R          : escape
 * - Ins/Del    : escape/ok
 * - Home/End   : ok/escape
 * <NumKeypad>
 * - 0/.        : shift/control
 * - 1/2/3      : ok/cancel/menu
 * - 4/6        : pageup/pagedown 
 * - 5          : escape
 * - 8          : tab
 * - +          : tab
 * - -          : escape
 * <Gamepad>
 * - LT         : tab
 * - RT         : control
 * - back       : escape
 * - start      : ok
 * 
 * -- Terms of Use --
 * Comply with the MIT license agreement.
 * http://www.opensource.org/licenses/mit-license
 * 
 */

/*:ja
 * @target MZ
 * @plugindesc 左右兼用のキー配置を適用します。
 * @author 水部
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_KeyPatchForBothLR.js
 * 
 * @help Sbm_KeyPatchForBothLR.js
 *
 * 【概要】
 * このプラグインは、左右兼用のキー配置を適用します。
 * 「WASDキーによる移動」、「テンキーでの基本操作」などの追加改変を含みます。
 * 左手で移動、右手で決定/キャンセル。もしくは左手のみでの操作が可能です。
 * 
 * 【使用方法】
 * 本プラグインを導入するだけで適用されます。
 * 
 * 【改変内容】
 * <キーボード>
 * ・W          ：pagedown  → up arrow
 * ・Alt        ：control   → alt
 * <テンキー>
 * ・0          ：escape    → shift
 * ・2          ：down      → cancel
 * ・4          ：left      → pageup
 * ・6          ：right     → pagedown
 * ・8          ：up        → tab
 * 
 * 【追加内容】
 * <キーボード>
 * ・W/A/S/D    ：up/left/down/right *(arrow)
 * ・Q/E        ：pageup/pagedown
 * ・C          ：ok
 * ・R          ：escape
 * ・Ins/Del    ：escape/ok
 * ・Home/End   ：ok/escape
 * <テンキー>
 * ・0/.        ：shift/control
 * ・1/2/3      ：ok/cancel/menu
 * ・4/6        ：pageup/pagedown
 * ・5          ：escape
 * ・8          ：tab
 * ・+          ：tab
 * ・-          ：escape
 * <ゲームパッド>
 * ・LT         ：tab
 * ・RT         ：control
 * ・back       ：escape
 * ・start      ：ok
 * 
 * 【利用規約】
 * MITライセンス規約に準じます。
 * https://licenses.opensource.jp/MIT/MIT
 * 
 */

(() => {
    'use strict';

    // ==============================
    // Input
    // ==============================
    const keyPatch = {
        18: "alt",          // alt
        45: "escape",       // insert
        46: "ok",           // delete
        36: "ok",           // home
        35: "escape",       // end
        67: "ok",           // C
        81: "pageup",       // Q
        69: "pagedown",     // E
        82: "escape",       // R
        65: "left",         // A
        68: "right",        // D
        83: "down",         // S
        87: "up",           // W
        96: "shift",        // numpad 0
        97: "ok",           // numpad 1
        98: "cancel",       // numpad 2
        99: "menu",         // numpad 3
        100: "pageup",      // numpad 4
        101: "escape",      // numpad 5
        102: "pagedown",    // numpad 6
        104: "tab",         // numpad 8
        107: "tab",         // numpad +
        109: "escape",      // numpad -
        110: "control",     // numpad .
    }
    Object.assign(Input.keyMapper, keyPatch);

    const gamepadPatch = {
        6: "tab",       // LT
        7: "control",   // RT
        8: "escape",    // back
        9: "ok",        // start
    }
    Object.assign(Input.gamepadMapper, gamepadPatch);
})();


Sbm_ReversibleKeyPatch.js
/*
===============================================================================
 RPG Maker MZ - Sbm ReversibleKeyPatch (1.0.0)
===============================================================================
 Copyright (c) 2021 Suib
 This software is released under the MIT License.
 http://www.opensource.org/licenses/mit-license
===============================================================================
*/

/*:
 * @target MZ
 * @plugindesc this plugin provides key patch that For both left and right uses.
 * @author Suib
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_ReversibleKeyPatch.js
 *
 * @help Sbm_ReversibleKeyPatch.js
 *
 * -- Summary --
 * this plugin provides key patch that For both left and right uses.
 * 
 * -- Usage --
 * It is applied only by installing this plugin.
 * 
 * -- Changes --
 * <Keyboard>
 * - W          : pagedown  -> up arrow
 * - Alt        : control   -> alt
 * <NumKeypad>
 * - 0          : escape    -> shift
 * - 2          : down      -> cancel
 * - 4          : left      -> pageup
 * - 6          : right     -> pagedown
 * - 8          : up        -> tab
 * 
 * -- Additions --
 * <Keyboard>
 * - W/A/S/D    : up/left/down/right *(arrow)
 * - Q/E        : pageup/pagedown
 * - C          : ok
 * - R          : escape
 * - Ins/Del    : escape/ok
 * - Home/End   : ok/escape
 * <NumKeypad>
 * - 0/.        : shift/control
 * - 1/2/3      : ok/cancel/menu
 * - 4/6        : pageup/pagedown 
 * - 5          : escape
 * - 8          : tab
 * - +          : tab
 * - -          : escape
 * <Gamepad>
 * - LT         : tab
 * - RT         : control
 * - back       : escape
 * - start      : ok
 * 
 * -- Terms of Use --
 * Comply with the MIT license agreement.
 * http://www.opensource.org/licenses/mit-license
 * 
 */

/*:ja
 * @target MZ
 * @plugindesc 左右兼用のキー配置を適用します。
 * @author 水部
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_ReversibleKeyPatch.js
 * 
 * @help Sbm_ReversibleKeyPatch.js
 *
 * 【概要】
 * このプラグインは、左右兼用のキー配置を適用します。
 * 「WASDキーによる移動」、「テンキーでの基本操作」などの追加改変を含みます。
 * 左手で移動、右手で決定/キャンセル。もしくは左手のみでの操作が可能です。
 * 
 * 【使用方法】
 * 本プラグインを導入するだけで適用されます。
 * 
 * 【改変内容】
 * <キーボード>
 * ・W          ：pagedown  → up arrow
 * ・Alt        ：control   → alt
 * <テンキー>
 * ・0          ：escape    → shift
 * ・2          ：down      → cancel
 * ・4          ：left      → pageup
 * ・6          ：right     → pagedown
 * ・8          ：up        → tab
 * 
 * 【追加内容】
 * <キーボード>
 * ・W/A/S/D    ：up/left/down/right *(arrow)
 * ・Q/E        ：pageup/pagedown
 * ・C          ：ok
 * ・R          ：escape
 * ・Ins/Del    ：escape/ok
 * ・Home/End   ：ok/escape
 * <テンキー>
 * ・0/.        ：shift/control
 * ・1/2/3      ：ok/cancel/menu
 * ・4/6        ：pageup/pagedown
 * ・5          ：escape
 * ・8          ：tab
 * ・+          ：tab
 * ・-          ：escape
 * <ゲームパッド>
 * ・LT         ：tab
 * ・RT         ：control
 * ・back       ：escape
 * ・start      ：ok
 * 
 * 【利用規約】
 * MITライセンス規約に準じます。
 * https://licenses.opensource.jp/MIT/MIT
 * 
 */

(() => {
    'use strict';

    // ==============================
    // Input
    // ==============================
    const keyPatch = {
        18: "alt",          // alt
        45: "escape",       // insert
        46: "ok",           // delete
        36: "ok",           // home
        35: "escape",       // end
        67: "ok",           // C
        81: "pageup",       // Q
        69: "pagedown",     // E
        82: "escape",       // R
        65: "left",         // A
        68: "right",        // D
        83: "down",         // S
        87: "up",           // W
        96: "shift",        // numpad 0
        97: "ok",           // numpad 1
        98: "cancel",       // numpad 2
        99: "menu",         // numpad 3
        100: "pageup",      // numpad 4
        101: "escape",      // numpad 5
        102: "pagedown",    // numpad 6
        104: "tab",         // numpad 8
        107: "tab",         // numpad +
        109: "escape",      // numpad -
        110: "control",     // numpad .
    }
    Object.assign(Input.keyMapper, keyPatch);

    const gamepadPatch = {
        6: "tab",       // LT
        7: "control",   // RT
        8: "escape",    // back
        9: "ok",        // start
    }
    Object.assign(Input.gamepadMapper, gamepadPatch);
})();


Sbm_KeyPatchForNumpad.js
/*
===============================================================================
 RPG Maker MZ - Sbm KeyPatchForNumpad (1.0.0)
===============================================================================
 Copyright (c) 2021 Suib
 This software is released under the MIT License.
 http://www.opensource.org/licenses/mit-license
===============================================================================
*/

/*:
 * @target MZ
 * @plugindesc This plugin assigns any key settings to the Numpad.
 * @author Suib
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_KeyPatchForNumpad.js
 * @base PluginCommonBase
 * @orderAfter Sbm_ReversibleKeyPatch
 *
 * @param keyMapArrNumber
 * @text keySettings:Number
 * @desc Assign operations to the key to "0-9" (multiple can be selected).
 * @type struct<KeyMapNumber>[]
 * @default []
 * 
 * @param keyMapArrSymbol
 * @text keySettings:Symbol
 * @desc Assign operations to keys such as "+" and "-" (multiple can be selected).
 * @type struct<KeyMapSymbol>[]
 * @default []
 * 
 * @help Sbm_KeyPatchForNumpad.js
 *
 * -- Summary --
 * This plugin assigns any key settings to the numeric keypad.
 * Unassigned keys retain their default key settings.
 * 
 * -- Usage --
 * Select the setting target from "Number" or "Symbol" in the plugin parameters.
 * Select any "keycode" in the pull-down item and select the "operation" to be assigned to it.
 * 
 * -- Plugin Parameters --
 * keySettings:Number (multiple can be selected)
 * keySettings:Symbol (multiple can be selected)
 * 
 * -- Base Plugins --
 * PluginCommonBase
 * 
 * -- Terms of Use --
 * Comply with the MIT license agreement.
 * http://www.opensource.org/licenses/mit-license
 * 
 */
/*~struct~KeyMapNumber:
 *
 * @param keyCode
 * @text keyCode
 * @desc Select the key to which you want to assign the operation.
 * @type select
 * @option Numpad 0
 * @value 96
 * @option Numpad 1
 * @value 97
 * @option Numpad 2
 * @value 98
 * @option Numpad 3
 * @value 99
 * @option Numpad 4
 * @value 100
 * @option Numpad 5
 * @value 101
 * @option Numpad 6
 * @value 102
 * @option Numpad 7
 * @value 103
 * @option Numpad 8
 * @value 104
 * @option Numpad 9
 * @value 105
 * 
 * @param operation
 * @text operation
 * @desc Select the operation to assign to the target key.
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */
/*~struct~KeyMapSymbol:
 *
 * @param keyCode
 * @text keyCode
 * @desc Select the key to which you want to assign the operation.
 * @type select
 * @option Numpad +
 * @value 107
 * @option Numpad -
 * @value 109
 * @option Numpad *
 * @value 106
 * @option Numpad /
 * @value 111
 * @option Numpad .
 * @value 110
 * 
 * @param operation
 * @text operation
 * @desc Select the operation to assign to the target key.
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */

/*:ja
 * @target MZ
 * @plugindesc このプラグインは、テンキーに任意のキー設定を割り当てます。
 * @author 水部
 * @url https://github.com/suibotsukuiki/RPGMakerMZ/blob/main/Sbm_KeyPatchForNumpad.js
 * @base PluginCommonBase
 * @orderAfter Sbm_ReversibleKeyPatch
 *
 * @param keyMapArrNumber
 * @text キー設定：数字(複数選択可)
 * @desc 「0-9」までのキーに操作を割り当てます(複数選択可)。
 * @type struct<KeyMapNumber>[]
 * @default []
 * 
 * @param keyMapArrSymbol
 * @text キー設定：記号(複数選択可)
 * @desc 「+」や「-」等のキーに操作を割り当てます(複数選択可)。
 * @type struct<KeyMapSymbol>[]
 * @default []
 * 
 * @help Sbm_KeyPatchForNumpad.js
 *
 * 【概要】
 * このプラグインは、テンキーに任意のキー設定を割り当てます。
 * 割り当てが行われなかったキーはデフォルトのキー設定を保持します。
 * 
 * 【使用方法】
 * ・プラグインパラメータで、設定対象のキーを「数字」または「記号」から選択します。
 * ・プルダウン項目で任意の「キー」を選択し、それに割り当てる「操作」を選択します。
 * 
 * 【プラグインパラメータ】
 * ・キー設定：数字(複数選択可)
 * ・キー設定：記号(複数選択可)
 * 
 * 【前提プラグイン】
 * ・PluginCommonBase
 * 
 * 【利用規約】
 * MITライセンス規約に準じます。
 * https://licenses.opensource.jp/MIT/MIT
 * 
 */
/*~struct~KeyMapNumber:ja
 *
 * @param keyCode
 * @text キーコード
 * @desc 操作を割り当てる対象のキーを選択します。
 * @type select
 * @option Numpad 0
 * @value 96
 * @option Numpad 1
 * @value 97
 * @option Numpad 2
 * @value 98
 * @option Numpad 3
 * @value 99
 * @option Numpad 4
 * @value 100
 * @option Numpad 5
 * @value 101
 * @option Numpad 6
 * @value 102
 * @option Numpad 7
 * @value 103
 * @option Numpad 8
 * @value 104
 * @option Numpad 9
 * @value 105
 * 
 * @param operation
 * @text 操作
 * @desc 対象のキーに割り当てる操作を選択します。
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */
/*~struct~KeyMapSymbol:ja
 *
 * @param keyCode
 * @text キーコード
 * @desc 操作を割り当てる対象のキーを選択します。
 * @type select
 * @option Numpad +
 * @value 107
 * @option Numpad -
 * @value 109
 * @option Numpad *
 * @value 106
 * @option Numpad /
 * @value 111
 * @option Numpad .
 * @value 110
 * 
 * @param operation
 * @text 操作
 * @desc 対象のキーに割り当てる操作を選択します。
 * @type select
 * @option ok
 * @option cancel
 * @option menu
 * @option escape
 * @option tab
 * @option shift
 * @option control
 * @option alt
 * @option pageup
 * @option pagedown
 * @option left
 * @option up
 * @option right
 * @option down
 * @option debug
 */

(() => {
    'use strict';
    const script = document.currentScript;
    const param = PluginManagerEx.createParameter(script);

    const keyMapNumpad = {};
    if (Array.isArray(param.keyMapArrNumber)) {
        for (const keyMap of param.keyMapArrNumber) {
            Object.assign(keyMapNumpad, { [String(keyMap.keyCode)]: keyMap.operation });
        }
    }
    if (Array.isArray(param.keyMapArrSymbol)) {
        for (const keyMap of param.keyMapArrSymbol) {
            Object.assign(keyMapNumpad, { [String(keyMap.keyCode)]: keyMap.operation });
        }
    }
    // ==============================
    // Input
    // ==============================
    Object.assign(Input.keyMapper, keyMapNumpad);
})();
